cluster-autoscaler/core/scale_up.go

```diff
@@ -475,36 +498,54 @@ func ScaleUp(context *context.AutoscalingContext, processors *ca_processors.Auto
 		return &status.ScaleUpStatus{
 				ScaledUp:                true,
 				ScaleUpInfos:            scaleUpInfos,
-				PodsRemainUnschedulable: getRemainingPods(podsRemainUnschedulable),
+				PodsRemainUnschedulable: getRemainingPods(podsRemainUnschedulable, skippedNodeGroups),
 				PodsTriggeredScaleUp:    bestOption.Pods,
 				PodsAwaitEvaluation:     getPodsAwaitingEvaluation(unschedulablePods, podsRemainUnschedulable, bestOption.Pods)},
 			nil
 	}
 
-	return &status.ScaleUpStatus{ScaledUp: false, PodsRemainUnschedulable: getRemainingPods(podsRemainUnschedulable)}, nil
+	return &status.ScaleUpStatus{ScaledUp: false, PodsRemainUnschedulable: getRemainingPods(podsRemainUnschedulable, skippedNodeGroups)}, nil
 }
 
-func getRemainingPods(podSet map[*apiv1.Pod]bool) []*apiv1.Pod {
-	result := make([]*apiv1.Pod, 0)
-	for pod, val := range podSet {
-		if val {
-			result = append(result, pod)
+func getRemainingPods(schedulingErrors map[*apiv1.Pod][]*simulator.PredicateError, skipped int) map[*apiv1.Pod]string {
+	remaining := map[*apiv1.Pod]string{}
+	for pod, errs := range schedulingErrors {
+		messages := []string{}
+		aggregated := map[string]int{}
+		for _, err := range errs {
+			// Aggregate by reason, since predicate name isn't as helpful.
+			// If two different predicates fail with the same reason, so be it.
+			for _, reason := range err.Reasons() {
```

Comment:
Scheduler aggregates by predicate name. I agree reason is probably more useful, but I wonder how well it aggregates? For example will we end with separate entries looking something like "insufficient cpu: required 2500m, available 2000m", "insufficient cpu: required 2500m, available 1000m?

I'd like to see some example events for most common scenarios (insufficient cpu/mem/gpu, taint, nodeSelector/nodeAffinity).

Reply:
I think basic resources are all checked by PodFitsResources predicate, and scheduler doesn't aggregate by this name.

Example event running current version of this PR for insufficient CPU:

```3m          23m          110       long-5786d47bf5-jr5ct.154d90743d10c944   Pod                      Normal    NotTriggerScaleUp   cluster-autoscaler      pod didn't trigger scale-up (it wouldn't fit if a new node is added): Insufficient cpu (1)```

It's accompanied by scheduler event for this pod:

```3m          23m          420       long-5786d47bf5-jr5ct.154d9071973d7470   Pod                      Warning   FailedScheduling    default-scheduler       0/3 nodes are available: 3 Insufficient cpu.```

Given that the scheduler's message format is now "<number of affected nodes> <failure reason>", we should probably use this as well. I'll change this.

Reply:
Fixed format to match scheduler's

Reply:
There's still a lot of room for improvement, e.g. we should probably make it clear that we're talking about node groups rejected for scale-up. Ideally there'd be a per-node-group reason, but then the message size would grow unreasonably long in some cases.

Reply:
Thanks for the example outputs. They look good for me. We can always improve formatting later if people find them hard to use, but I think using the same format as scheduler provides a consistent experience.